lib/demo_web/controllers/user_reset_password_controller.ex

```diff
@@ -0,0 +1,59 @@
+defmodule DemoWeb.UserResetPasswordController do
+  use DemoWeb, :controller
+
+  alias Demo.Accounts
+
+  plug :get_user_by_reset_password_token when action in [:edit, :update]
+
+  def new(conn, _params) do
+    render(conn, "new.html")
+  end
+
+  def create(conn, %{"user" => %{"email" => email}}) do
+    if user = Accounts.get_user_by_email(email) do
+      Accounts.deliver_user_reset_password_instructions(
+        user,
+        &Routes.user_reset_password_url(conn, :edit, &1)
+      )
+    end
+
+    # Regardless of the outcome, show an impartial success/error message.
+    conn
+    |> put_flash(
+      :info,
+      "If your e-mail is in our system, you will receive instructions to reset your password shortly."
+    )
+    |> redirect(to: "/")
+  end
+
+  def edit(conn, _params) do
+    render(conn, "edit.html", changeset: Accounts.change_user_password(conn.assigns.user))
+  end
+
+  # Do not login the user after reset password to avoid a
+  # leaked token giving the user access to the account.
```

Comment:
Would you mind elaborating here a little bit?
My thought process was that if the attacker reset the user password, they already have access to the account so it doesn't matter if we log the user in.

There's a case where I can see it being better not to log the user in, for example, if the attacker intercepts the token but doesn't know the user email. Is this to prevent this case?

Reply:
There are multiple ways a token can leak. For example, it can leak in logs or URL referrers. In those, you generally don't know the e-mail. However, if it leaks by e-mail, then you are right, it is game over.

Reply:
Thanks for the clarification!

Reply:
Please notice one thing: the token leaking through the e-mail is not the only way how a leak of both the e-mail and token may happen.

1. E.g. It may leak in server logs (then to our hosted log manager etc.) when we log to whom we send e-mails.
2. E.g. Somebody could also target an individual and knowing his e-mail address, and observing him and noticing the moment when he requests and e-mail reset, and having access to e.g. server or proxy logs etc. – by using some automation – login before the user would be able to use the token. Yes, this is hard to do, but still possible.

To mitigate this it is needed to don't send to the user a password reset link that has the token, but send him the link and the token below with an instruction to visit the link and paste in the code.
Or better after the user requests a password reset redirect him to the password reset page and put in the form a "Reset token" field and some message asking him to put here the reset token send to his e-mail address. It is done on some sites I have seen (e.g. Microsoft https://account.live.com/).